source/common/quic/envoy_quic_session_cache.cc

```diff
@@ -0,0 +1,168 @@
+#include "common/quic/envoy_quic_session_cache.h"
+
+namespace Envoy {
+namespace Quic {
+namespace {
+
+constexpr size_t MaxSessionCacheEntries = 1024;
+
+// Returns false if the SSL |session| doesn't exist or it is expired at |now|.
+bool isSessionValid(SSL_SESSION* session, SystemTime now) {
+  if (session == nullptr) {
+    return false;
+  }
+  const time_t now_time_t = std::chrono::system_clock::to_time_t(now);
+  if (now_time_t < 0) {
+    return false;
+  }
+  const uint64_t now_u64 = static_cast<uint64_t>(now_time_t);
+  const uint64_t session_time = SSL_SESSION_get_time(session);
+  const uint64_t session_expiration = session_time + SSL_SESSION_get_timeout(session);
+  // now_u64 may be slightly behind because of differences in how time is calculated at this layer
+  // versus BoringSSL. Add a second of wiggle room to account for this.
+  return session_time <= now_u64 + 1 && now_u64 < session_expiration;
+}
+
+bool doApplicationStatesMatch(const quic::ApplicationState* state, quic::ApplicationState* other) {
+  if (state == other) {
+    return true;
+  }
```

Comment:
Yes, this is needed to avoid a dereference when they're both nullptr. I personally prefer this style of comparison because it is correct for all inputs. The fact that callers won't ever send the same non-null pointer as both parameters is an implementation details that doesn't need to be visible in this function.